File: Python/131. Palindrome Partitioning.py
# 131. Palindrome Partitioning.

# Topic: String, Dynamic Programming, Backtracking.


'''
Task:
-----
Given a string s, partition s such that every substring  of the partition is a palindrome. 
Return all possible palindrome partitioning of s.

Example 1:
Input: s = "aab"
Output: [["a","a","b"],["aa","b"]]

Example 2:
Input: s = "a"
Output: [["a"]]
 
Constraints:
1 <= s.length <= 16
s contains only lowercase English letters.

# Testcase:
------------
"aab"
"a"

# Code:
-------------
class Solution:
    def partition(self, s: str) -> List[List[str]]:

'''
# Solution:
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)

        # Precompute all palindrome substrings in the string iteratively.
        # is_palindrome[i][j] will be True if the substring s[i:j+1] is a palindrome.
        is_palindrome = [[False] * n for _ in range(n)]
        for i in range(n):
            is_palindrome[i][i] = True  # A single character is always a palindrome.
            for j in range(i):
                # A substring s[j:i+1] is a palindrome if the characters at the ends are equal,
                # and if the substring s[j+1:i] is a palindrome.
                if s[j] == s[i] and (i - j == 1 or is_palindrome[j + 1][i - 1]):
                    is_palindrome[j][i] = True

        # Backtracking function to build all palindrome partitions.
        def backtrack(start, path):
            # If the start index reaches the end of the string, add the current path to results.
            if start == n:
                result.append(path[:])
                return

            # Explore all possible partitions by expanding the end index.
            for end in range(start, n):
                # If the substring s[start:end+1] is a palindrome, continue backtracking.
                if is_palindrome[start][end]:
                    backtrack(end + 1, path + [s[start:end + 1]])

        result = []
        backtrack(0, [])
        return result

# Test cases
solution = Solution()
print(solution.partition("aab"))  # Example 1
print(solution.partition("a"))    # Example 2



# Description:
'''
To solve the problem of splitting a string into palindromes,we can optimize the algorithm 
by minimizing the number of recursive calls and palindrome checks. One way to achieve this
is to use an iterative approach to pre-generate all palindromic substrings, and then apply 
the backtracking method to compile partitions based on this data.

An iterative search approach for all palindromes will reduce the total number of checks performed 
during the backtracking process.

In this code:

- We first create a 2D list is_palindrome to store whether each substring is a palindrome. This is done iteratively.
- In the backtrack function, we explore all possible partitions of the string. If a substring is a palindrome 
  (as checked using is_palindrome), we continue exploring further partitions from the end of this substring.
- The function backtrack is initially called with the start index 0 and an empty path. It recursively builds up paths
  of palindromes and adds them to the result list when a complete partition of the string is found.
- The function returns all the possible palindrome partitioning of the input string s.

'''

File: Cpp/131. Palindrome Partitioning.cpp
// 131. Palindrome Partitioning.

//Topic: String, Dynamic Programming, Backtracking.

/*
### Task:
---
Given a string s, partition s such that every substring of the partition is a palindrome. 
Return all possible palindrome partitioning of s.

Example 1:
Input: s = "aab"
Output: [["a","a","b"],["aa","b"]]

Example 2:
Input: s = "a"
Output: [["a"]]

Constraints:
1 <= s.length <= 16
s contains only lowercase English letters.


### Testcase:
---
"aab"
"a"


### Code:
---
class Solution {
public:
    vector<vector<string>> partition(string s) {
        
    }
};

*/
// Solution:  ------------------------------------------------------------------------------

class Solution {
public:
    // Function to check if a string is a palindrome.
    // This function uses dynamic programming to store results of palindrome checks.
    bool isPalindrome(const string &s, int start, int end, vector<vector<int>> &dp) {
        if (start >= end) return true;
        if (dp[start][end] != -1) return dp[start][end];

        dp[start][end] = (s[start] == s[end]) && isPalindrome(s, start + 1, end - 1, dp);
        return dp[start][end];
    }

    // Backtracking function to find all palindrome partitions.
    void backtrack(int start, const string &s, vector<string> &path, vector<vector<string>> &result, vector<vector<int>> &dp) {
        if (start >= s.size()) {
            result.push_back(path);
            return;
        }

        for (int end = start; end < s.size(); end++) {
            if (isPalindrome(s, start, end, dp)) {
                path.push_back(s.substr(start, end - start + 1));
                backtrack(end + 1, s, path, result, dp);
                path.pop_back(); // Backtrack
            }
        }
    }

    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        vector<string> path;
        vector<vector<int>> dp(s.size(), vector<int>(s.size(), -1)); // Dynamic programming table

        backtrack(0, s, path, result, dp);
        return result;
    }
};





// Description:  ============================================================================
/*
To solve the "Palindrome Partitioning" problem, we can utilize backtracking combined with dynamic programming. 
The goal is to find all the ways we can split a given string such that every substring is a palindrome.

### Approach:
1. **Backtracking**: We use backtracking to explore all possible partitioning of the string.
2. **Check for Palindromes**: For each partitioning, we check if the current substring is a palindrome.
3. **Dynamic Programming for Palindrome Checking**: Use dynamic programming to store the results of palindrome checks 
     for substrings, to avoid redundant calculations.
4. **Partitioning**: Recursively create different partitions and backtrack if the current partition is not a palindrome.

### Explanation:
- `isPalindrome` function checks if a substring is a palindrome. It uses a DP table (`dp`) to store the results of previous 
   checks to avoid redundant calculations.
- `backtrack` function explores all possible partitions of the string. For each partition, it checks if it forms a palindrome. 
   If yes, it proceeds to partition the remaining string.
- In `backtrack`, `path` holds the current partition, and `result` accumulates all valid palindrome partitions.
- `dp` table in `partition` function is initialized with `-1`, indicating that no palindrome checks have been done yet.
- The main function `partition` initializes necessary structures and starts the backtracking process.

### Complexity Analysis:
- **Time Complexity**: O(N * 2^N), where N is the length of the string. In the worst case, we might end up checking every 
    substring for being a palindrome and explore all partitioning.
- **Space Complexity**: O(N^2) for the DP table plus O(N) for the recursion stack, leading to O(N^2) overall. The DP table 
    stores palindrome checks for every possible substring.

*/